bpf: dynamic-width static data inlining

ti-mo opened the pull request:
```
Historically, the largest data type used for declaring global data was uint32.
Cilium's iproute2 fork had patches that made the assumption all values are
32 bits, so this behaviour was replicated when it was ported to Go. This only
works well if the largest variable emitted to .data is 32 bits wide. Any
smaller and the section's alignment would drop, causing the inliner to read
part of the next variable. Any larger and only part of a u64 would be read.

Since we now no longer rely on iproute2 for loading any ELFs, we can start
changing the way this works. In Go, we have access to .data's BTF Datasec,
so we can be a bit smarter about it and detect how wide each variable is,
so we can accurately copy out the bits needed.

This patch introduces a uint64 global data macro, DEFINE_U64, and updates
DEFINE_IPV6 to use it. This cuts down on the amount of instructions needed
to carry ipv6 values in the instruction stream from 8 to 4.
```

Closes: #7123

```release-note
Reduce amount of bpf instructions needed for handling ipv6 addresses
```

ti-mo committed:
bpf/lib: sync endian.h with upstream

Missing bpf_cpu_to_be64 for follow-up commit.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: dynamic-width static data inlining

Historically, the largest data type used for declaring global data was uint32.
Cilium's iproute2 fork had patches that made the assumption all values are
32 bits, so this behaviour was replicated when it was ported to Go. This only
works well if the largest variable emitted to .data is 32 bits wide. Any
smaller and the section's alignment would drop, causing the inliner to read
part of the next variable. Any larger and only part of a u64 would be read.

Since we now no longer rely on iproute2 for loading any ELFs, we can start
changing the way this works. In Go, we have access to .data's BTF Datasec,
so we can be a bit smarter about it and detect how wide each variable is,
so we can accurately copy out the bits needed.

This patch introduces a uint64 global data macro, DEFINE_U64, and updates
DEFINE_IPV6 to use it. This cuts down on the amount of instructions needed
to carry ipv6 values in the instruction stream from 8 to 4.

Signed-off-by: Timo Beckers <timo@isovalent.com>

gentoo-root commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -211,7 +211,7 @@ func (s *ConfigSuite) TestWriteStaticData(c *C) {
-		for _, suffix := range []string{"_1", "_2", "_3", "_4"} {
+		for _, suffix := range []string{"_1", "_2"} {
 			// Variables with these suffixes are implemented via
 			// multiple 32-bit values. The header define doesn't
```
Nit: since this patch it can be either 32-bit or 64-bit.

gentoo-root commented on pkg/datapath/loader/template.go:
```diff
@@ -216,10 +216,8 @@ func elfVariableSubstitutions(ep datapath.Endpoint) map[string]uint32 {
-		result["LXC_IP_3"] = sliceToBe32(ipv6[8:12])
-		result["LXC_IP_4"] = sliceToBe32(ipv6[12:16])
+		result["LXC_IP_1"] = sliceToBe32(ipv6[0:8])
+		result["LXC_IP_2"] = sliceToBe32(ipv6[8:16])
```
Why Be32 if it's 64-bit now? This won't work, will it?

ti-mo replied:
Good catch, was a little rushed writing this. This made me realize the ELF substitution code also needed 64-bit support added.

gentoo-root commented on pkg/datapath/loader/template.go:
```diff
@@ -216,10 +216,8 @@ func elfVariableSubstitutions(ep datapath.Endpoint) map[string]uint32 {
-		result["LXC_IP_2"] = sliceToBe32(ipv6[4:8])
-		result["LXC_IP_3"] = sliceToBe32(ipv6[8:12])
-		result["LXC_IP_4"] = sliceToBe32(ipv6[12:16])
+		result["LXC_IP_1"] = sliceToBe32(ipv6[0:8])
```
General question: why isn't the same mechanism used for NAT_46X64_PREFIX? Can we still optimize NAT_46X64_PREFIX as well to use two 64-bits instead of four 32-bits?

ti-mo replied:
Yep, the way it's implemented looks a bit strange to me. Instead of assigning individual bytes, why don't we write a single value to `v6addr->p1`, which is a u32?  @borkmann

ti-mo replied:
I figure it might be related to endianness and making sure the same byte goes into the same slot in the address. The way `sliceToBe32` and friends functions is quite subtle, and the ELF rewriter also potentially byte-swaps when converting u16/32/64 to []byte, which bit me a few times while working on this PR. Some of it will go away when we remove ELF rewriting, though, I'll revisit it then.

ti-mo committed:
datapath,elf: support 64-bit ELF substitutions

This patch adds support for 64-bit wide ELF substitutions, to be used in
a follow-up patch.

Signed-off-by: Timo Beckers <timo@isovalent.com>

lmb commented on pkg/bpf/collection.go:
```diff
@@ -237,11 +238,11 @@ func classifyProgramTypes(spec *ebpf.CollectionSpec) error {
 // emit direct array accesses instead of memory loads with an offset from the
 // map's pointer.
 func inlineGlobalData(spec *ebpf.CollectionSpec) error {
```
Interesting, I was wondering whether something like this existed!

Nit: doc needs updating.

> Currently, all map reads are expected to be 32 bits wide until BTF MapKV

lmb commented on pkg/bpf/collection_test.go:
```diff
@@ -63,9 +64,17 @@ func TestInlineGlobalData(t *testing.T) {
 			globalDataMap: {
+				Value: &btf.Datasec{
+					Vars: []btf.VarSecinfo{
+						{Offset: 0, Size: 4},
```
Nit: add an entry with padding here?

ti-mo replied:
Good idea, changed var 2 to be 2 bytes wide at offset 4 and blinded the padding with 0xff to make sure the inliner ignores it.

pchaigno reviewed (approved):
LGTM. Thanks for the nice commit history!

joestringer reviewed (approved):
API codeowners changes are trivial, approving on their behalf.

CI is green, ready to merge.
